Bluetooth: Add support for changing Device Name

Vudentz opened the pull request:
This adds support for changing GAP Device Name at runtime and persists it in a flash if BT_SETTINGS is defined.

Note: It is also possible to set BT_DEVICE_NAME_STORAGE without BT_SETTINGS but that will be overwritten in the next boot thus why there is a line with default 0 if !BT_SETTINGS.

Fixes #8357

Vudentz committed:
Bluetooth: Store device name

This uses bt_dev to store the name and allow changing it at runtime, in
addtion to that if CONFIG_BT_SETTINGS is defined make the name
persistent.

Fixes #8357

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Make GAP name writable

This makes GAP name writable if CONFIG_BT_DEVICE_NAME_STORAGE is > 0
which means the name can be persisted.

Fixes #8357

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Add BT_LE_ADV_OPT_USE_NAME

This introduces a new advertising flag BT_LE_ADV_OPT_USE_NAME which can
be used by applications to make the stack automatically include the
Bluetooth Device Name in the Scan Response.

The name is also updated in case there is already an advertising
instance using it.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: samples: Make use of BT_LE_ADV_OPT_USE_NAME

Use BT_LE_ADV_OPT_USE_NAME whenver possible since it does properly
track updates.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: peripheral: Set CONFIG_BT_DEVICE_NAME_MAX

This force CONFIG_BT_DEVICE_NAME_MAX to be set even in boards where
CONFIG_BT_SETTINGS won't be defined so it exercise such configuration.

In addition to that set a long name that would not fix in the
ScanData.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -335,6 +335,17 @@ config BT_SCAN_WITH_IDENTITY
 	  disclosing local identity information. However, if the use case
 	  requires disclosing it then enable this option.
 
+config BT_DEVICE_NAME_STORAGE
```
I'd call it `BT_DEVICE_NAME_MAX` since we haven't (at least so far) had the convention of using `STORAGE` for the upper limit of something.

Vudentz replied:
Ok

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -335,6 +335,17 @@ config BT_SCAN_WITH_IDENTITY
 	  requires disclosing it then enable this option.
 
+config BT_DEVICE_NAME_STORAGE
+	int "Size in bytes for device name"
```
*Maximum* size in bytes..

Vudentz replied:
Ok

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -335,6 +335,17 @@ config BT_SCAN_WITH_IDENTITY
+	help
+	  Bluetooth device name storage size. Storage can be up to 248 bytes
+	  long. 0 means that no storage space is allocated for device name in
+	  which case the device name cannot be changed at runtime.
```
You should make it clear here that this includes the terminating null character.

Vudentz replied:
Right, I hadn't mentioned it is a string though but I guess one can still assume so it is better to make it clear here.

jhedberg commented on subsys/bluetooth/host/Kconfig:
```diff
@@ -335,6 +335,17 @@ config BT_SCAN_WITH_IDENTITY
+config BT_DEVICE_NAME_STORAGE
+	int "Size in bytes for device name"
+	default BT_RX_BUF_LEN
+	default BT_L2CAP_RX_MTU if BT_HCI_ACL_FLOW_CONTROL
```
These defaults seem a bit odd to me. I'd make the default something that will always fit in full in the scan response data, meaning 28 (if I calculated correctly), or then what will always fit within the default ATT MTU.

Vudentz replied:
Right, I was thinking only in terms of ATT MTU but since we may also advertise that should be the default. We may also need to truncate and use the short time in the advertisement if the name is set to be bigger than that.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4694,6 +4698,92 @@ int bt_enable(bt_ready_cb_t cb)
+	size_t len = strlen(name);
+	int err;
+
+	if (len > sizeof(bt_dev.name)) {
```
Shouldn't this be >= since bt_dev.name should include the terminating null?

Vudentz replied:
Right.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4694,6 +4698,92 @@ int bt_enable(bt_ready_cb_t cb)
+	int err;
+
+	if (len > sizeof(bt_dev.name)) {
+		return -EINVAL;
```
Wouldn't -ENOMEM be more appropriate?

Vudentz replied:
Right.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4694,6 +4698,92 @@ int bt_enable(bt_ready_cb_t cb)
+	}
+
+	memset(bt_dev.name, 0, sizeof(bt_dev.name));
+	strncpy(bt_dev.name, name, sizeof(bt_dev.name) - 1);
```
If you fix the length check I commented on earlier, I think the memset may become unnecessary, and you should then remove the `- 1` from the above line.

Vudentz replied:
Yep.

jhedberg commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -4802,6 +4859,21 @@ int bt_le_adv_start(const struct bt_le_adv_param *param,
+		/* Cannot use name if sd is set */
+		if (sd) {
+			return -EINVAL;
+		}
```
Can't we do better than this? I think it would be friendlier if we'd just append the user-given SD elements after the name. I think there may well be valid use cases for other elements in the SD while still wanting to use the configured name there as well.

Vudentz replied:
Hmm, like replacing the SD with another array and then making that point to user SD? I guess that could work but what I was thinking is that we would need to check for the name to be part of the SD already in which case we would have to fail.

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -166,6 +178,15 @@ static void generate_static_addr(void)
 }
 
+#if CONFIG_BT_DEVICE_NAME_STORAGE > 0
+static void generate_name(void)
```
This isn't really generating anything, so the name seems wrong. That said, the function is super short and you're only calling it from one place, so I'd just do the bt_set_name() in line from commit(). Extra bonus is that you get less ugly `#ifdefs` in this file that way as well.

Vudentz replied:
Sounds good, that was a bigger function initially but I endup redoing it as bt_set_name.

Vudentz committed:
Bluetooth: Allow use of ScanData with BT_LE_ADV_OPT_USE_NAME

This enables the user to provide a ScanData, as long as it contain
names, and set BT_LE_ADV_OPT_USE_NAME.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Use shortened name if complete doesn't fit

If the complete name doesn't fit try shortening the name.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: shell: Make use of BT_LE_ADV_OPT_USE_NAME

Use BT_LE_ADV_OPT_USE_NAME whenver possible since it does properly
track updates.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: shell: Add name command

name command can be used to read or write the GAP Device Name which is
used by the advertise command.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -46,18 +46,48 @@ struct ccc_store {
+	}
+
+	if (offset > sizeof(value)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
```
Wouldn't it make sense to return this error for any non-zero offset? If you copy to a non-zero offset that you end up with an empty string (since the first byte will be zero), which makes little sense and would likely anyway be different from what the GATT client expects. It would also simplify the code below.

Vudentz replied:
Yep, seems this should behave as a string it makes sense to not allow non-zero offset.

jhedberg commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -46,18 +46,48 @@ struct ccc_store {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (offset + len > sizeof(value)) {
```
Shouldn't this be >=? Otherwise you risk not having a null byte at the end.

Vudentz replied:
Right, will fix that.

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -121,6 +121,18 @@ static int set(int argc, char **argv, char *val)
 	}
 
+	if (!strcmp(argv[0], "name")) {
+		len = sizeof(bt_dev.name);
```
Usage of "len" seems unnecessary ere since sizeof(bt_dev.name) is a compile time constant (i.e. you can just use that everywhere you use len at the moment)

Vudentz replied:
Done.

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -121,6 +121,18 @@ static int set(int argc, char **argv, char *val)
+			       " will be truncated");
+		}
+
+		strncpy(bt_dev.name, val, len);
```
The last parameter should be "len - 1" (or "sizeof(bt_dev.name) - 1") since otherwise you risk leaving it without a terminating null.

Vudentz replied:
Done.

jhedberg commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -649,13 +662,10 @@ static int cmd_advertise(int argc, char *argv[])
-		scan_rsp = sd;
-		scan_rsp_len = ARRAY_SIZE(sd);
+		param.options = BT_LE_ADV_OPT_CONNECTABLE |
+				BT_LE_ADV_OPT_USE_NAME;
```
I'd put parenthesis around the bitwise operation. I think that's the general convention, checkpatch may complain otherwise, and it'll also make editors (at least mine) align the second line correctly :)

Vudentz replied:
Done

jhedberg reviewed (changes requested):
A few more things

jhedberg commented on subsys/bluetooth/host/settings.c:
```diff
@@ -121,6 +121,17 @@ static int set(int argc, char **argv, char *val)
+			       " will be truncated");
+		}
+
+		strncpy(bt_dev.name, val, sizeof(bt_dev.name) - 1);
```
Shouldn't this be without the `-1`? Otherwise, if you get `strlen(val) == sizeof(bt_dev.name) - 1`  the result of strncpy will not be null terminated (unless `bt_dev.name[sizeof(bt_dev.name - 1)]` has been initialized to zero at some point).

Vudentz replied:
Aren't we restoring the value after boot here? I though that would be initialized with 0, note that we still would have to prevent non-null terminated when strlen >= sizeof since that just print a warning and truncates the string, but perhaps in that case we can just set the last byte to '\0' since we know it would be truncated.

jli157 commented:
@Vudentz How can I change the device name from the central device side other than using the function `bt_set_name`?
